File: ferromagnetism/moodle/magnetfit2.py
"""Python translation of magnetfit2.m."""
import numpy as np
from scipy.optimize import minimize_scalar

KB = 1.38e-23
MU = 0.928e-23

def magnetfit(I, Ierr, T, Tc, H, Isat):
    """Return (chi2, Msol) for the mean-field magnetization fit.

    Msol is normalized by the saturation value.
    """
    I = np.asarray(I, dtype=float)
    Ierr = np.asarray(Ierr, dtype=float)
    T = np.asarray(T, dtype=float)

    Msol = np.zeros_like(T)
    for k, Tk in enumerate(T):
        res = minimize_scalar(
            lambda m: abs(m - np.tanh(MU * H / (KB * Tk) + (Tc / Tk) * m)),
            bounds=(0, 1), method='bounded',
        )
        Msol[k] = res.x

    chi2 = np.sum((I / Isat - Msol) ** 2 / (Ierr / Isat) ** 2)
    return chi2, Msol

File: ferromagnetism/moodle/magnetfit2.m
function [chi2,Msol] = magnetfit(I,Ierr,T,Tc,H,Isat)

% Please notice this function returns 2 outputs! MATLAB requires a vector [a,b] to save them ( [a,b]=magnetfit(I,Ierr,T,Tc,H,Isat) )
% Also notice, the theoretical output "Msol" is already normalized by the saturation value.
% Convince yourself that inputting current/voltage/field will yield the same result since we are always normalizing by saturation 
% This function computes the chi2 value for a fit of the normalized 
% magnetization curve
% Input:  I    = Measured induced current
%         Ierr = Error of the measured current
%         T    = Temperature of the material
%         Tc   = Curie's temparature
%         H    = External field
%         Isat = The saturation value of the current
% Output: chi2 = chi square value
%         Msol = The model of the magnatization curve


kb=1.38e-23;    % Boltzman's constant [Joul/Kelvin]
mu = 0.928e-23; % Magnetic moment [Joul/Tesla]

Msol = zeros(size(T));
for K=1:length(T);
    % re-write the magnetization equation
    mf = inline(['abs(m-tanh(' num2str(mu) '*' num2str(H) '/(' num2str(kb) '*' num2str(T(K)) ')+(' num2str(Tc) '/' num2str(T(K)) ')*m))']);
    Msol(K)=fminbnd(mf,0,1);  % find the zero of the function, 
                              % i.e., for a given Tc,H, and T what is m    
end


% compute chi2
chi2 = sum((I/Isat - Msol).^2 ./ (Ierr/Isat).^2);